frontend/src/metabase/lib/data_grid.js

```diff
@@ -28,93 +28,76 @@ export function multiLevelPivot(
     valuesByKey[valueKey] = valueColumnIndexes.map(index => row[index]);
   }
 
-  const [headerFormatters, valueFormatters, rowHeaderFormatters] = [
-    columnColumnIndexes,
-    valueColumnIndexes,
-    rowColumnIndexes,
-  ].map(indexes =>
-    indexes.map(index => value =>
-      formatValue(value, { column: data.cols[index] }),
-    ),
+  const valueFormatters = valueColumnIndexes.map(index => value =>
+    formatValue(value, { column: data.cols[index] }),
   );
-  const headerRows = getHeaderRows(columnColumnTree, {
-    valueColumns: valueColumnIndexes.map(index => data.cols[index]),
-    formatters: headerFormatters,
-  });
 
-  const bodyRows = getBodyRows(rowColumnTree, {
-    columnValueLists: valueLists(columnColumnTree),
-    valuesByKey,
-    valueColumnCount: valueColumnIndexes.length,
-    rowHeaderFormatters,
-    valueFormatters,
-  });
+  const valueColumns = valueColumnIndexes.map(index => data.cols[index]);
 
   return {
-    headerRows,
-    bodyRows,
+    topIndex: getIndex(columnColumnTree, { valueColumns }),
+    leftIndex: getIndex(rowColumnTree, {}),
+    getRowSection: createRowSectionGetter({
+      valuesByKey,
+      columnColumnTree,
+      rowColumnTree,
+      valueFormatters,
+    }),
   };
 }
 
-function getHeaderRows(values, { valueColumns, formatters }) {
-  if (values.length === 0) {
-    // if we have multiple value columns include their column names
-    return valueColumns.length > 1
-      ? [valueColumns.map(c => ({ value: c.display_name, span: 1 }))]
-      : [];
-  }
-  const [currentFormatter, ...otherFormatters] = formatters;
-  const rowLists = [];
-  const currentRow = values.map(({ value, children }) => {
-    const rows = getHeaderRows(children, {
-      valueColumns,
-      formatters: otherFormatters,
-    });
-    rowLists.push(rows);
-    const span =
-      rows.length === 0 ? 1 : rows[0].reduce((sum, { span }) => sum + span, 0);
-    return { value: currentFormatter(value), span };
-  });
-  const followingRows = _.zip(...rowLists).map(a => a.flat());
-  return [currentRow, ...followingRows];
+function createRowSectionGetter({
+  valuesByKey,
+  columnColumnTree,
+  rowColumnTree,
+  valueFormatters,
+}) {
+  return (topValue, leftValue) => {
+    const rows = enumerate(
+      rowColumnTree.find(node => node.value === leftValue),
+    );
+    const columns = enumerate(
+      columnColumnTree.find(node => node.value === topValue),
+    );
+    return rows.map(row =>
+      columns.flatMap(col => {
+        const valueKey = JSON.stringify(col.concat(row));
+        const values = valuesByKey[valueKey];
+        if (values === undefined) {
+          return new Array(valueFormatters).fill(null);
```

Comment:
alternative: `return Array(valueFormatters.length).fill(null)`
(assuming here only the length is needed, since the contents are replaced by null anyway)

Reply:
Whoops good catch! The code I had would always result in `[null]` I think.